File: skills/together-gpu-clusters/scripts/manage_cluster.ts
#!/usr/bin/env -S npx tsx
/**
 * Together AI GPU Clusters -- Create, Monitor, Scale, Delete
 *
 * Full lifecycle: list regions, create cluster, wait for ready,
 * check status, scale, then delete.
 *
 * Usage:
 *   npx tsx manage_cluster.ts
 *
 * Requires:
 *   npm install together-ai
 *   export TOGETHER_API_KEY=your_key
 */

import Together from "together-ai";

const client = new Together({
  apiKey: process.env.TOGETHER_API_KEY,
});

async function listRegions(): Promise<void> {
  console.log("=== Available Regions ===");
  const regions = await client.beta.clusters.list_regions();
  for (const r of regions.regions) {
    console.log(`  ${r.name}: GPUs=${JSON.stringify(r.supported_instance_types)}`);
  }
}

async function listClusters(): Promise<void> {
  console.log("\n=== Existing Clusters ===");
  const response = await client.beta.clusters.list();
  for (const c of response.clusters) {
    console.log(`  ${c.cluster_id}: ${c.cluster_name} (${c.status}, ${c.num_gpus} GPUs)`);
  }
}

async function createCluster(
  name: string,
  region: string,
  gpuType: string,
  numGpus: number,
  driverVersion: string,
  billingType: string = "ON_DEMAND",
  clusterType: string = "KUBERNETES",
): Promise<any> {
  const cluster = await client.beta.clusters.create({
    cluster_name: name,
    region,
    gpu_type: gpuType,
    num_gpus: numGpus,
    driver_version: driverVersion,
    billing_type: billingType,
    cluster_type: clusterType,
  });
  console.log(`Created cluster: ${cluster.cluster_id}  (status: ${cluster.status})`);
  return cluster;
}

async function waitForReady(
  clusterId: string,
  timeoutMs: number = 1_800_000,
  pollMs: number = 30_000,
): Promise<any> {
  const start = Date.now();
  while (Date.now() - start < timeoutMs) {
    const cluster = await client.beta.clusters.retrieve(clusterId);
    const elapsed = Math.round((Date.now() - start) / 1000);
    console.log(`  Status: ${cluster.status}  (${elapsed}s)`);

    if (cluster.status === "Ready") return cluster;
    if (cluster.status === "Deleting") {
      throw new Error(`Cluster is being deleted: ${clusterId}`);
    }

    await new Promise((r) => setTimeout(r, pollMs));
  }
  throw new Error(`Cluster not ready after ${timeoutMs / 1000}s`);
}

async function scaleCluster(clusterId: string, numGpus: number): Promise<any> {
  const cluster = await client.beta.clusters.update({
    cluster_id: clusterId,
    num_gpus: numGpus,
  });
  console.log(`Scaled to ${numGpus} GPUs (status: ${cluster.status})`);
  return cluster;
}

async function deleteCluster(clusterId: string): Promise<void> {
  await client.beta.clusters.delete(clusterId);
  console.log(`Deleted cluster: ${clusterId}`);
}

async function main(): Promise<void> {
  const CLUSTER_NAME = "my-training-cluster";
  const REGION = "us-central-8";
  const GPU_TYPE = "H100_SXM";
  const NUM_GPUS = 8;
  const DRIVER = "CUDA_12_6_560";

  // 1. List available regions
  await listRegions();

  // 2. List existing clusters
  await listClusters();

  // 3. Create a cluster
  const cluster = await createCluster(
    CLUSTER_NAME, REGION, GPU_TYPE, NUM_GPUS, DRIVER,
  );

  // 4. Wait for cluster to be ready
  console.log("\nWaiting for cluster to be ready...");
  const ready = await waitForReady(cluster.cluster_id);
  console.log(`Cluster ready: ${ready.cluster_name}`);

  // 5. Scale up to 16 GPUs
  console.log("\nScaling to 16 GPUs...");
  await scaleCluster(cluster.cluster_id, 16);

  // 6. Wait for scaling to complete
  await waitForReady(cluster.cluster_id);

  // 7. Delete when done (uncomment to delete)
  // await deleteCluster(cluster.cluster_id);
}

main();

File: skills/together-gpu-clusters/scripts/manage_cluster.py
#!/usr/bin/env python3
"""
Together AI GPU Clusters -- Create, Monitor, Scale, Delete (v2 SDK)

Full lifecycle: list regions, create cluster, wait for ready,
check status, scale, then delete.

Usage:
    python manage_cluster.py

Requires:
    pip install together
    export TOGETHER_API_KEY=your_key
"""

import time
from together import Together

client = Together()


def list_regions():
    """List available regions with supported GPUs and drivers."""
    regions = client.beta.clusters.list_regions()
    for r in regions.regions:
        print(f"  {r.name}: GPUs={r.supported_instance_types}, "
              f"Drivers={r.driver_versions}")
    return regions


def list_clusters():
    """List all GPU clusters."""
    response = client.beta.clusters.list()
    for c in response.clusters:
        print(f"  {c.cluster_id}: {c.cluster_name} "
              f"({c.status}, {c.num_gpus} GPUs, {c.gpu_type})")
    return response.clusters


def create_cluster(
    name: str,
    region: str,
    gpu_type: str,
    num_gpus: int,
    driver_version: str,
    billing_type: str = "ON_DEMAND",
    cluster_type: str = "KUBERNETES",
    volume_id: str | None = None,
):
    """Create a new GPU cluster."""
    kwargs: dict = {
        "cluster_name": name,
        "region": region,
        "gpu_type": gpu_type,
        "num_gpus": num_gpus,
        "driver_version": driver_version,
        "billing_type": billing_type,
        "cluster_type": cluster_type,
    }
    if volume_id:
        kwargs["volume_id"] = volume_id

    cluster = client.beta.clusters.create(**kwargs)
    print(f"Created cluster: {cluster.cluster_id}  (status: {cluster.status})")
    return cluster


def wait_for_ready(cluster_id: str, timeout: int = 1800, poll_interval: int = 30):
    """Poll until cluster reaches Ready state."""
    elapsed = 0
    while elapsed < timeout:
        cluster = client.beta.clusters.retrieve(cluster_id)
        print(f"  Status: {cluster.status}  ({elapsed}s)")

        if cluster.status == "Ready":
            return cluster
        if cluster.status in ("Deleting",):
            raise RuntimeError(f"Cluster is being deleted: {cluster_id}")

        time.sleep(poll_interval)
        elapsed += poll_interval

    raise TimeoutError(f"Cluster not ready after {timeout}s")


def scale_cluster(cluster_id: str, num_gpus: int):
    """Scale a cluster to a new GPU count (must be multiple of 8)."""
    cluster = client.beta.clusters.update(cluster_id, num_gpus=num_gpus)
    print(f"Scaled cluster {cluster_id} to {num_gpus} GPUs (status: {cluster.status})")
    return cluster


def delete_cluster(cluster_id: str):
    """Delete a GPU cluster."""
    client.beta.clusters.delete(cluster_id)
    print(f"Deleted cluster: {cluster_id}")


if __name__ == "__main__":
    CLUSTER_NAME = "my-training-cluster"
    REGION = "us-central-8"
    GPU_TYPE = "H100_SXM"
    NUM_GPUS = 8
    DRIVER = "CUDA_12_6_560"

    # 1. List available regions
    print("Available regions:")
    list_regions()

    # 2. List existing clusters
    print("\nExisting clusters:")
    list_clusters()

    # 3. Create a cluster
    cluster = create_cluster(
        name=CLUSTER_NAME,
        region=REGION,
        gpu_type=GPU_TYPE,
        num_gpus=NUM_GPUS,
        driver_version=DRIVER,
    )

    # 4. Wait for cluster to be ready
    print("\nWaiting for cluster to be ready...")
    cluster = wait_for_ready(cluster.cluster_id)
    print(f"Cluster ready: {cluster.cluster_name}")

    # 5. Scale up to 16 GPUs
    print("\nScaling to 16 GPUs...")
    scale_cluster(cluster.cluster_id, 16)

    # 6. Wait for scaling to complete
    cluster = wait_for_ready(cluster.cluster_id)

    # 7. Delete when done (uncomment to delete)
    # delete_cluster(cluster.cluster_id)
